### ISSUE TITLE
Performance gains by ONNX inconsistent

### ISSUE THREAD

**Issue by riqiang-dp** [microsoft/onnxruntime#12348]
Hi, I'm just hoping to get some attention to this question, if there are any answers, if this is expected behavior, or if this indicates some underlying issue that could be fixed. Original post below by someone else, and I'm seeing very similar results with slightly different setups.

Macbook Pro:
CPU: Intel i7-9750H (12) @ 2.60GHz
Mem: 16G

VM (Google Cloud VM, N1-standard-1):
CPU: Xeon E5-2699 v3 @ 2.30GHz, 45MB cache (Intel Haswell)
Mem: 7.5G

Inference with one thread, one process. The numbers are similar to what the orignal poster stated, where ONNX is 1/3 the delay compared to Pytorch on the Macbook, but about the same on the VM. The baseline Pytorch performance for the VM and the Macbook is similar however for my usecase, which makes it more frustrating to see the huge speed boost on Macbook but the performance lags behind on VM. I'm wondering if this due to onnxruntime or Pytorch being unoptimized on some of the platforms. 

### Discussed in https://github.com/microsoft/onnxruntime/discussions/11100

<div type='discussions-op-text'>

<sup>Originally posted by **Kroshtan** April  4, 2022</sup>
Please inform me if this is the wrong place to ask.

I have a BERT based model, exported to ONNX and I am trying to gauge the performance increase. For GPU, the performance makes sense, and it excellent, but I am also trying to check performance on CPU. 

The setup is as follows: A Bert encoder with a single linear layer and softmax layer attached, created and trained in pytorch. Exported with pytorch to onnx. Using ONNX runtime to run the model. Additionally, the model is optimized using `onnxruntime/transformers/optimize.py`.

These are the values I have gotten so far, leaving all other variables constant:

Framework | Location | Time/iteration (single-threaded in brackets)
--- | --- | ---
Pytorch | CPU, laptop local | 250ms (1510ms)
ONNX | CPU, laptop local | 130ms (420ms)
ONNX (onnxruntime/transformers/optimize.py) | CPU, laptop local | 124ms (409ms)
Pytorch | GPU, laptop local | 15ms
ONNX | GPU, laptop local |  
Pytorch | CPU, VM | 150ms (671ms)
ONNX | CPU, VM | 240ms (606ms)
ONNX (onnxruntime/transformers/optimizer.py) | CPU, VM | 222ms (609ms)
Pytorch | GPU, VM (V100) | 15ms
ONNX | GPU, VM (V100) | 10ms
ONNX (onnxruntime/transformers/optimizer.py) | GPU, VM (V100) | 3.20ms

VM hardware:
Intel(R) Xeon(R) CPU E5-2690 v4 @ 2.60GHz
Thread(s) per core:  1
Core(s) per socket:  6
Socket(s):                1
GPU: Tesla V100

Local hardware:
AMD Ryzen 7 4800H with Radeon Graphics
Thread(s) per core:              2
Core(s) per socket:              8
Socket(s):                       1
GPU: NVidia GeForce RTX 3050

Now here's the issue:
Why is the performance increase between PyTorch and ONNX locally so much bigger than on the VM? Obviously there is a difference between the local hardware and the VM, and a difference between multithreaded and single-threaded, which is exacerbated by the choice of hardware, but I do not understand why the difference between PyTorch and ONNX is not consistent, within the same environment.
</div>

**Comment by tianleiwu** [comment]
@riqiang-dp, could you share your onnx model so that we can take a look?  If you do not want to share weights, you can exclude them from onnx model in python like the following:
```
import onnx
model = onnx.load("input.onnx")
onnx.save_model(model, "output.onnx", save_as_external_data=True)
```
The output.onnx will be very small to share.

You can also try our benchmark tool on bert-base-cased model:
https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/python/tools/transformers/run_benchmark.sh
The performance could be related to whether the onnx model is fully optimized or not etc, model size (like number of heads, hidden size), input size (batch size and sequence length) and cpu threads and the hardware itself.

**Comment by riqiang-dp** [comment]
Thanks for the reply, I'll try exporting the model as well as the benchmark tool.

**Comment by riqiang-dp** [comment]
Hi I've exported the encoder model without data: https://drive.google.com/file/d/1ybOsfFF7tx5yORAgj-SAP4c_DV6Cx0dr/view?usp=sharing
Are you able to take a look?

I'vee also tried the benchmarking tool on our VM:
<img width="1294" alt="image" src="https://user-images.githubusercontent.com/58701255/185212382-6535f0a8-c69b-4cb4-b19f-d185d0a0c00c.png">
It seems like there should at least be some improvement using onnxruntime with fp32, but for quantized int8 models torchscript and onnxruntime works similarly? However I was comparing onnxruntime to Pytorch ckpt files so there should still be an improvement?

Otherwise I have tried the optimization script [here](https://github.com/microsoft/OLive/tree/master/cmd-example), and different runs gave different results. The conclusion I eventually just set num_threads to 1, no changes to other things.
